src/apps/chifra/cmd/serve/exec/admin_pins.go

```diff
@@ -0,0 +1,62 @@
+package serveRoutes
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/cmd/pins"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/pinlib"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/pinlib/manifest"
+)
+
+var ErrMultipleModes = errors.New("only one of ?list or ?init is supported")
+var ErrNoMode = errors.New("you must choose one of ?list or ?init")
+
+// AdminPins runs pin-operating functions depending on query parameters
+func AdminPins(request *http.Request) ([]manifest.PinDescriptor, error) {
+	query := request.URL.Query()
```

Comment:
Couldn't we convert this into a `cobra` command and/or convert the cobra command into an http.Request and/or convert both of them into a key/value map so the validation and execution of the commands could be the same code? Perhaps you already anticipate that?

Reply:
I totally get your point and I agree. But there are some problems to solve first:
1. Cobra's `Command.Run` returns nothing, so how do we get the results?
2. Running as a server requires a slight change in the behaviour of the programme. For example, we shouldn't use `log.Fatal` in the command, because this would kill the whole server. 
3. Running as server switches the code from printing out results and errors to returning them (this is why I separated HandleList into List and HandleList. The latter prints output, the former does the job and returns data as variables)

Reply:
There is a version of `cobra's` `Run` called `RunE` which returns an error. I've been thinking we may want to use that instead of `Run`. I think people use it to do 'extra' validation before running but perhaps after doing some processing. So, validate is for things that can be validated early. `RunE` is for stuff that can't be validated until later.